kernel/virtio/virtio_blk.c

```diff
@@ -44,176 +44,84 @@ struct virtio_blk_hdr {
-    struct virtio_blk_req *req;
-    int dbg = 0;
-
-    if (((blk_next_avail + 3) % blkq.num)
-        == ((blk_last_used * 3) % blkq.num)) {
-        printf("blk buffer full! next_avail:%d last_used:%d\n",
-               blk_next_avail, blk_last_used);
-        return NULL;
-    }
-
-    /* we perform a copy into the blk buffer to make reclaiming easy */
-    assert(len <= VIRTIO_BLK_SECTOR_SIZE);
-    req = &blk_bufs[blk_next_avail/3];
-
-    if (req->hw_used) {
-        printf("blk buffer full! no unused buffers!\n");
-        return NULL;
-    }
-
-    req->hdr.type = type;
-    req->hdr.ioprio = 0;
-    req->hdr.sector = sector;
-    req->hw_used = 0;
-
-    if (type == VIRTIO_BLK_T_OUT)
-        memcpy(req->data, data, len);
-
-    /* it seems like we need 3 descriptors for a single req */
-
-    if (dbg)
-        atomic_printf("REQ BLK: 0x%p type=%d\n", req, req->hdr.type);
-
-
-    /* the header */
-    desc = &(blkq.desc[blk_next_avail]);
-    desc->addr = (uint64_t)&req->hdr;
-    desc->len = sizeof(struct virtio_blk_hdr);
-    desc->next = (blk_next_avail + 1) % blkq.num;
-    desc->flags = VIRTQ_DESC_F_NEXT;
-
-    if (dbg)
-        atomic_printf("REQ BLK: hdr at 0x%x\n", desc->addr);
-
-    /* the data */
-    desc = &(blkq.desc[(blk_next_avail + 1) % blkq.num]);
-    desc->addr = (uint64_t)req->data;
-    desc->len = len;
-    desc->next = (blk_next_avail + 2) % blkq.num;
-    desc->flags = VIRTQ_DESC_F_NEXT;
-    if (type == VIRTIO_BLK_T_IN)
-        desc->flags |= VIRTQ_DESC_F_WRITE;
-
-    if (dbg)
-        atomic_printf("REQ BLK: data at 0x%x data=%08x...\n",
-                      desc->addr, *(uint32_t *)desc->addr);
-
-    /* the status */
-    desc = &(blkq.desc[(blk_next_avail + 2) % blkq.num]);
-    desc->addr = (uint64_t)&req->status;
-    desc->len = sizeof(uint8_t);
-    desc->next = 0;
-    desc->flags = VIRTQ_DESC_F_WRITE;
-
-    if (dbg)
-        atomic_printf("REQ BLK: status at 0x%x\n", desc->addr);
+    uint16_t head;
+    struct io_buffer *head_buf, *data_buf, *status_buf;
+    int ret = -1;
 
-    if (dbg)
-        atomic_printf("BLK: 0x%p next_avail %d last_used %d\n",
-                      desc->addr, blk_next_avail, blk_last_used);
+    head = virtio_blk_op(type, sector, data, *len);
 
-    avail = blkq.avail;
-    /* Memory barriers should be unnecessary with one processor */
-    blkq.avail->ring[avail->idx % blkq.num] = blk_next_avail;
+    head_buf = &blkq.bufs[head];
+    data_buf = &blkq.bufs[(head + 1) % blkq.num];
+    status_buf = &blkq.bufs[(head + 2) % blkq.num];
 
-    avail->idx++;
-    blk_next_avail = (blk_next_avail + 3) % blkq.num;
-    outw(virtio_blk_pci_base + VIRTIO_PCI_QUEUE_NOTIFY, VIRTQ_BLK);
+    /* XXX need timeout or something, because this can hang... sync
+     * should probably go away anyway
+     */
+    while (!head_buf->completed)
+        ;
 
-    return req;
-}
+    ret = (*(uint8_t *) status_buf) != VIRTIO_BLK_S_OK;
+    if (type == VIRTIO_BLK_T_OUT && ret == 0)
```

Comment:
It felt because there is a bug in the test (https://github.com/Solo5/solo5/blob/master/tests/test_blk/test_blk.c#L22):

```
        if (sector_write[i] != '0' + i % 10) <== should be sector_read
            /* Check failed */
            return 1;
```